python/cudf/cudf/core/udf/groupby_utils.py

```diff
@@ -0,0 +1,181 @@
+# Copyright (c) 2022-2023, NVIDIA CORPORATION.
+
+import math
+import os
+
+import cupy as cp
+import numpy as np
+from numba import cuda, types
+from numba.cuda.cudadrv.devices import get_context
+from numba.np import numpy_support
+from numba.types import Record
+
+import cudf.core.udf.utils
+from cudf.core.udf.groupby_typing import (
+    SUPPORTED_GROUPBY_NUMPY_TYPES,
+    Group,
+    GroupType,
+)
+from cudf.core.udf.templates import (
+    group_initializer_template,
+    groupby_apply_kernel_template,
+)
+from cudf.core.udf.utils import (
+    _get_kernel,
+    _get_ptx_file,
+    _get_udf_return_type,
+    _supported_cols_from_frame,
+    _supported_dtypes_from_frame,
+)
+from cudf.utils.utils import _cudf_nvtx_annotate
+
+dev_func_ptx = _get_ptx_file(os.path.dirname(__file__), "function_")
+cudf.core.udf.utils.ptx_files.append(dev_func_ptx)
+
+
+def _get_frame_groupby_type(dtype, index_dtype):
+    """
+    Get the numba `Record` type corresponding to a frame.
+    Models the column as a dictionary like data structure
+    containing GroupTypes.
+    See numba.np.numpy_support.from_struct_dtype for details.
+    """
+
+    # Create the numpy structured type corresponding to the numpy dtype.
+    fields = []
+    offset = 0
+
+    sizes = [val[0].itemsize for val in dtype.fields.values()]
+    for i, (name, info) in enumerate(dtype.fields.items()):
+        elemdtype = info[0]
+        title = info[2] if len(info) == 3 else None
+        ty = numpy_support.from_dtype(elemdtype)
+        indexty = numpy_support.from_dtype(index_dtype)
+        infos = {
+            "type": GroupType(ty, indexty),
+            "offset": offset,
+            "title": title,
+        }
+        fields.append((name, infos))
+
+        offset += 8 + 8 + 8  # group struct size (2 pointers and 1 integer)
+
+        # Align the next member of the struct to be a multiple of the
+        # memory access size, per PTX ISA 7.4/5.4.5
+        if i < len(sizes) - 1:
+            # next_itemsize = sizes[i + 1]
+            next_itemsize = 8
+            offset = int(math.ceil(offset / next_itemsize) * next_itemsize)
+
+    # Numba requires that structures are aligned for the CUDA target
+    _is_aligned_struct = True
+    return Record(fields, offset, _is_aligned_struct)
+
+
+def _groupby_apply_kernel_string_from_template(frame, args):
+    """
+    Function to write numba kernels for `Groupby.apply` as a string.
+    Workaround until numba supports functions that use `*args`
```

Comment:
Is there a bug/feature request in for this? @gmarkall ?

Reply:
https://github.com/numba/numba/issues/6891